openpype/hosts/maya/api/lib.py

```diff
@@ -3176,38 +3243,78 @@ def _colormanage(**kwargs):
 def parent_nodes(nodes, parent=None):
     # type: (list, str) -> list
     """Context manager to un-parent provided nodes and return them back."""
-    import pymel.core as pm  # noqa
 
-    parent_node = None
-    delete_parent = False
+    def _as_mdagpath(node):
+        """Return MDagPath for node path."""
+        if not node:
+            return
+        sel = OpenMaya.MSelectionList()
+        sel.add(node)
+        return sel.getDagPath(0)
+
+    # We can only parent dag nodes so we ensure input contains only dag nodes
+    nodes = cmds.ls(nodes, type="dagNode", long=True)
+    if not nodes:
+        # opt-out early
+        yield
+        return
 
+    parent_node_path = None
+    delete_parent = False
     if parent:
         if not cmds.objExists(parent):
-            parent_node = pm.createNode("transform", n=parent, ss=False)
+            parent_node = cmds.createNode("transform",
+                                          name=parent,
+                                          skipSelect=False)
             delete_parent = True
         else:
-            parent_node = pm.PyNode(parent)
+            parent_node = parent
+        parent_node_path = cmds.ls(parent_node, long=True)[0]
+
+    # Store original parents
     node_parents = []
     for node in nodes:
-        n = pm.PyNode(node)
-        try:
-            root = pm.listRelatives(n, parent=1)[0]
-        except IndexError:
-            root = None
-        node_parents.append((n, root))
+        node_parent = get_node_parent(node)
+        node_parents.append((_as_mdagpath(node), _as_mdagpath(node_parent)))
+
     try:
-        for node in node_parents:
-            if not parent:
-                node[0].setParent(world=True)
+        for node, node_parent in node_parents:
```

Comment:
out of scope for this PR but couldn't we combine both `for` loops into one:
```
node_parents = []
try:
      for node in nodes:
              node_parent = get_node_parent(node)
              node_dag = _as_mdagpath(node)
              node_parent_dag = _as_mdagpath(node_parent)
              node_parents.append((node_dag, node_parent_dag))
              node_parent_path = node_parent_dag.fullPathName() if node_parent_dag else None # noqa

...
```

Reply:
In this case, no. In the second for loop we're reparenting nodes which means we're changing paths of maya nodes. It could be that one of the input nodes later in the list is a child node of one of the earlier nodes in the list. As such, once we get to that node in the iteration the node path has become invalid. That's why we're precollecting their `MFnDagNode` api equivalents prior to that before we're making any changes in the scene.

Or did I misunderstand?